client/windows/CCastleInterface.cpp

```diff
@@ -804,10 +804,25 @@ void CCastleBuildings::enterBlacksmith(ArtifactID artifactID)
 		LOCPLINT->showInfoDialog(boost::str(boost::format(CGI->generaltexth->allTexts[273]) % town->town->buildings.find(BuildingID::BLACKSMITH)->second->getNameTranslated()));
 		return;
 	}
-	auto art = artifactID.toArtifact(CGI->artifacts());
+	auto art = dynamic_cast<const CArtifact*>(artifactID.toArtifact(CGI->artifacts()));
 
 	int price = art->getPrice();
-	bool possible = LOCPLINT->cb->getResourceAmount(Res::GOLD) >= price && !hero->hasArt(artifactID);
+	bool possible = LOCPLINT->cb->getResourceAmount(Res::GOLD) >= price;
+	if(possible)
+	{
+		for(auto slot : art->possibleSlots.at(ArtBearer::HERO))

```

Comment:
what if possibleSlots.at will return empty container?

Reply:
@Nordsoft91 is this not handled in range-based for specification? I think it possibleSlots.begin() == possibleSlots.end() cycle will not run at all.

Reply:
> what if possibleSlots.at will return empty container?

If the artifact has no possible slots at all, then it can only be placed in a backpack. In this case, the purchase will be allowed.